### DISCUSSION TITLE
Changing keyboard layouts dynamically

### DISCUSSION THREAD

**Discussion by milisarge** [CuarzoSoftware/Louvre#62] [Support] [answered]
Is it possible to change keyboard layout dynamically without restarting the compositor?

**Comment by Fox2Code** [comment]
Translation: Is there an easy to use API to change keyboard layout at runtime.

Also, yeah, I think keyboard layout switching is important, I have a French keyboard, and sometimes I have to switch keyboard layout around to help me make some games work correctly.

**Comment by ehopperdietzel** [comment, marked as answer]
> Is it possible to change keyboard layout dynamically without restarting the compositor?

Yes, [LKeyboard::setKeymap()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_keyboard.html#af3224ac0e4648b52dbb8da1281b83549) can be called at any time and will update all clients keymaps.

**Comment by ehopperdietzel** [comment]
Hi @Fox2Code @milisarge, I've removed the hardcoded `setKeymap()` call inside `LCompositor::initialize()`. Now, the default keyboard map can be set using [XKB environment variables](https://cuarzosoftware.github.io/Louvre/environment_page.html).

**Reply by Fox2Code** [reply]
This is great stuff, assuming XKB won't give us an invalid initial layout.

**Reply by ehopperdietzel** [reply]
😮 You're right, when the envs are set, XKB doesn't return a fallback keymap on failure. I just tested unsetting them (on failure), and now it works again. Glad you noticed it!

**Reply by Fox2Code** [reply]
Well, glad I was able to help!
